### PR TITLE
Fix #147: Add semanticVersion property

### LINKED ISSUE #147: Add semanticVersion property. Require that direct producers produce it.
Converters should stick to populating 'version' which is the unmodified version of the tool.

### PR DESCRIPTION
**Pull request by ghost**
@michaelcfanning

### CODE REVIEW

File: document/Bibliography.mdk [hunk 1]
```diff
@@ -17,6 +17,3 @@ Available from &lt;<https://tools.ietf.org/html/rfc3629>&gt;
 
 RFC 3986, Uniform Resource Identifier (URI): Generic Syntax.
 Available from &lt;<https://tools.ietf.org/html/rfc3986>&gt;
-
-Semantic Versioning 2.0.0 \[viewed 2015-06-02\]. Available from &lt;<http://semver.org>&gt;
```
Comment (ghost) [inline document/Bibliography.mdk:21]: Moved to "Normative" since it's now part of a "shall" requirement.

File: document/CoreFormat.mdk [hunk 2]
```diff
@@ -753,9 +787,9 @@ a `formattedMessage` object (&sect;[#formattedMessage]) that can be used to cons
 a formatted message that describes the result.
 
 If the `formattedMessage` property is present on a `result`, the `fullMessage` property (&sect;[#result-fullMessage])
-shall not be present.
+shall be absent.
```
Comment (ghost) [inline document/CoreFormat.mdk:790]: Unrelated change. Everywhere else in the spec I say "shall be absent". These were the only two places I said "shall not be present."

File: document/CoreFormat.mdk [hunk 3]
```diff
@@ -413,25 +414,58 @@ along with its version and any other useful identifying information, such as its
...
+pattern specified by SemVer.
 ~
 
+### `version` property { #tool-version }
+
+In a log file produced by a static analysis tool, a `tool` object may contain a property named `version`
+whose value is a string containing the tool version in whatever format the tool natively provides.
+
+In a log file produced by a converter, the `version` property shall be present.
+
```
Comment (ghost) [inline document/CoreFormat.mdk:462]: We talked about having consumers attempt to interpret `version` as a SemVer version if `semanticVersion` was absent. But I don't think they should do that (so I didn't write that text here). If `semanticVersion` is absent, it means that the file's producer did not know whether to interpret the tool's native version string format as a semantic version. So I don't think consumers should make that guess either. I mean, they're welcome to try, but I wouldn't have the spec push them in that direction.

(If you disagree, let's talk offline and I'll capture the result as a reply to this comment. No need to have the whole discussion here.)

File: document/CoreFormat.mdk [hunk 4]
```diff
@@ -390,7 +390,8 @@ the post-processing tool should not alter any part of the `tool` object.
 {
     "name": "CodeScanner",                                       # see \(&sect;[#tool-name]\)
     "fullName": "CodeScanner 2.0, Developer Preview (en-US)",    # see \(&sect;[#tool-fullName]\)
-    "version": "2.0.0-beta",                                     # see \(&sect;[#tool-version]\)
+    "semanticVersion": "2.0.0-beta.1",                           # see \(&sect;[#tool-semanticVersion]\)
+    "version": "2.0 Beta 1 x86 (en-US)                           # see \(&sect;[#tool-version]\)
```
Comment (michaelcfanning) [inline document/CoreFormat.mdk]: The point of version, I thought, was to accurately reflect the literal version (and only version) provided by the tool. Your 'version' example looks like the full name (contains build flavor, locale, etc).

For a C# version, this would be the AssemblyVersion, for example (which isn't semver 2 compatible due to the Revision component).

File: document/CoreFormat.mdk [hunk 5]
```diff
@@ -413,25 +414,58 @@ along with its version and any other useful identifying information, such as its
...
 Semantic versions have the property of being sortable in chronological order of release.
+The presence of the `semanticVersion` property allows results management systems to (for example)
+restrict the results they display to versions newer than a specified version,
+or to restrict the results to a particular major version.
 ~
 
-~ IsoExample
-`"1.0.0-beta"`
+If the tool does not natively present its version string in SemVer format,
+it shall synthesize a SemVer string to populate the `semanticVersion` property.
```
Comment (michaelcfanning) [inline document/CoreFormat.mdk]: should you say something along the lines of 'otherwise, it should recap its native version string in semanticVersion'. although you said 'shall' earlier, it's worth restating here, do you think? also, should we mention the converter case? i.e., don't synthesize

File: document/CoreFormat.mdk [hunk 6]
```diff
@@ -413,25 +414,58 @@ along with its version and any other useful identifying information, such as its
...
+(the "pre-release" information is not in SemVer format).
+The tool would synthesize a SemVer string `"2.0.0-beta.1"`.
+~
+
+In a log file produced by a conversion tool, the `semanticVersion` property shall be absent.
+
+~ IsoNumberedNote
+The rationale is that a static analysis tool knows whether its version string is intended to be
+interpreted according to SemVer.
+A converter will in general not know this, even if the tool's version string conforms to the
```
Comment (michaelcfanning) [inline document/CoreFormat.mdk:452]: Can we sneak in the principle here that converters should not synthesize information, as a rule?

File: document/CoreFormat.mdk (hunk 3 above)
Comment (michaelcfanning) [inline document/CoreFormat.mdk:462]: no, makes sense.

**Comment by michaelcfanning** [comment]
LGTM. I do think you should fix up the 'version' example which doesn't really reflect the sort of 'native' versioning scheme we have in mind.

File: document/CoreFormat.mdk (hunk 6 above)
Comment (ghost) [inline document/CoreFormat.mdk:452]: Rather than sneak it in on this particular instance, I'd like to promote that to some top-level guidance at the beginning of the spec. I filed #156, "Add guidance for direct producers vs. converters".

File: document/CoreFormat.mdk (hunk 4 above)
Comment (ghost) [inline document/CoreFormat.mdk]: Changed.

Comment (ghost) [inline document/CoreFormat.mdk]: After some research on this, changed to use a "real" example of non-numeric versioning: Apple's major-minor-bug-stage-prerelease schema, _e.g._, `1.1.2b12`, mapping to `1.1.2-beta.12`.
